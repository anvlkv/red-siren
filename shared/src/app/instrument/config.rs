use serde::{Deserialize, Serialize};

#[derive(Debug, Default, Serialize, Deserialize, Clone, PartialEq)]
pub struct Config {
    pub portrait: bool,
    pub width: f32,
    pub height: f32,
    pub breadth: f32,
    pub length: f32,
    pub groups: usize,
    pub buttons_group: usize,
    pub button_size: f32,
}

impl Eq for Config {}

impl Config {
    pub fn new(width: f32, height: f32) -> Self {
        let portrait = height > width;

        let (length, breadth) = if portrait {
            (height, width / 3.0)
        } else {
            (width, height / 3.0)
        };

        let bs_size_b = breadth * 0.6;
        let bs_size_l = 932.0 / 6.0 * length / 932.0;

        let button_size = bs_size_b.min(bs_size_l);

        let max_buttons = (length / button_size).floor() as usize;

        let slots = max_buttons.div_euclid(2);

        let groups_2 = slots.div_euclid(2);
        let groups_3 = slots.div_euclid(3);
        let groups_5 = slots.div_euclid(5);

        let (groups, buttons_group) = if groups_5 > 1 {
            (groups_5, 5)
        } else if groups_3 > 1 {
            (groups_3, 3)
        } else {
            (groups_2, 2)
        };

        Config {
            portrait,
            width,
            height,
            length,
            breadth,
            button_size,
            groups: groups.try_into().unwrap_or(1),
            buttons_group,
        }
    }
}
