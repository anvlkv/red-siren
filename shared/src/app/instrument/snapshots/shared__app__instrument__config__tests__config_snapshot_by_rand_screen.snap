---
source: shared/src/app/instrument/config.rs
expression: configs
---
- portrait: false
  width: 1920
  height: 1080
  breadth: 360
  length: 1248
  whitespace: 306
  groups: 1
  buttons_group: 3
  button_size: 153
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 165.93439
- portrait: false
  width: 2560
  height: 1440
  breadth: 480
  length: 1796
  whitespace: 352
  groups: 1
  buttons_group: 3
  button_size: 176
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 133.21661
- portrait: false
  width: 1366
  height: 768
  breadth: 256
  length: 990
  whitespace: 158
  groups: 1
  buttons_group: 3
  button_size: 79
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 133.10132
- portrait: false
  width: 2732
  height: 2048
  breadth: 682.6666666666666
  length: 2016
  whitespace: 328
  groups: 2
  buttons_group: 2
  button_size: 164
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 167.5174
- portrait: false
  width: 2436
  height: 1125
  breadth: 375
  length: 2124
  whitespace: 126
  groups: 3
  buttons_group: 3
  button_size: 63
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 157.63445
- portrait: false
  width: 2960
  height: 1440
  breadth: 480
  length: 2740
  whitespace: 80
  groups: 6
  buttons_group: 3
  button_size: 40
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 140.76036
- portrait: false
  width: 3840
  height: 2160
  breadth: 720
  length: 2836
  whitespace: 472
  groups: 2
  buttons_group: 2
  button_size: 236
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 216.21936
- portrait: false
  width: 1280
  height: 800
  breadth: 266.6666666666667
  length: 840
  whitespace: 190
  groups: 1
  buttons_group: 3
  button_size: 95
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 161.66632
- portrait: false
  width: 3440
  height: 1440
  breadth: 480
  length: 2720
  whitespace: 330
  groups: 1
  buttons_group: 5
  button_size: 165
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 114.92094
- portrait: false
  width: 2560
  height: 1600
  breadth: 533.3333333333334
  length: 2036
  whitespace: 232
  groups: 1
  buttons_group: 3
  button_size: 116
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 181.54999
- portrait: true
  width: 1080
  height: 2340
  breadth: 360
  length: 2151
  whitespace: 72
  groups: 4
  buttons_group: 2
  button_size: 36
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 152.25151
- portrait: false
  width: 6016
  height: 3384
  breadth: 1128
  length: 4836
  whitespace: 560
  groups: 1
  buttons_group: 3
  button_size: 280
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 160.33766
- portrait: false
  width: 2048
  height: 1536
  breadth: 512
  length: 1492
  whitespace: 248
  groups: 2
  buttons_group: 2
  button_size: 124
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 111.70553
- portrait: false
  width: 2960
  height: 1440
  breadth: 480
  length: 2740
  whitespace: 80
  groups: 6
  buttons_group: 3
  button_size: 40
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 134.94023
- portrait: false
  width: 1280
  height: 720
  breadth: 240
  length: 996
  whitespace: 112
  groups: 2
  buttons_group: 2
  button_size: 56
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 207.54759
- portrait: false
  width: 2560
  height: 1440
  breadth: 480
  length: 2036
  whitespace: 232
  groups: 1
  buttons_group: 3
  button_size: 116
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 162.16452
- portrait: false
  width: 1280
  height: 1024
  breadth: 341.3333333333333
  length: 636
  whitespace: 292
  groups: 1
  buttons_group: 2
  button_size: 146
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 208.71033
- portrait: false
  width: 3840
  height: 1080
  breadth: 360
  length: 3312
  whitespace: 234
  groups: 4
  buttons_group: 2
  button_size: 117
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 125.59783
- portrait: false
  width: 2224
  height: 1668
  breadth: 556
  length: 1444
  whitespace: 360
  groups: 1
  buttons_group: 3
  button_size: 180
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 205.73105
- portrait: false
  width: 2960
  height: 1440
  breadth: 480
  length: 2740
  whitespace: 80
  groups: 6
  buttons_group: 3
  button_size: 40
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 141.00795
- portrait: false
  width: 3840
  height: 1600
  breadth: 533.3333333333334
  length: 3228
  whitespace: 276
  groups: 2
  buttons_group: 3
  button_size: 138
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 125.6121
- portrait: false
  width: 1280
  height: 720
  breadth: 240
  length: 996
  whitespace: 112
  groups: 2
  buttons_group: 2
  button_size: 56
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 114.66763
- portrait: false
  width: 1920
  height: 1200
  breadth: 400
  length: 1480
  whitespace: 190
  groups: 2
  buttons_group: 2
  button_size: 95
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 120.23089
- portrait: false
  width: 2560
  height: 1440
  breadth: 480
  length: 2148
  whitespace: 176
  groups: 1
  buttons_group: 3
  button_size: 88
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 150.8244
- portrait: false
  width: 1366
  height: 768
  breadth: 256
  length: 910
  whitespace: 198
  groups: 1
  buttons_group: 3
  button_size: 99
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 119.04943
- portrait: false
  width: 2560
  height: 1440
  breadth: 480
  length: 2380
  whitespace: 60
  groups: 3
  buttons_group: 3
  button_size: 30
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 133.21661
- portrait: false
  width: 1280
  height: 800
  breadth: 266.6666666666667
  length: 836
  whitespace: 192
  groups: 1
  buttons_group: 3
  button_size: 96
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 147.08772
- portrait: false
  width: 2960
  height: 1440
  breadth: 480
  length: 2740
  whitespace: 80
  groups: 6
  buttons_group: 3
  button_size: 40
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 134.94023
- portrait: false
  width: 3840
  height: 2160
  breadth: 720
  length: 2836
  whitespace: 472
  groups: 2
  buttons_group: 2
  button_size: 236
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 120.944435
- portrait: false
  width: 2560
  height: 1600
  breadth: 533.3333333333334
  length: 2036
  whitespace: 232
  groups: 1
  buttons_group: 3
  button_size: 116
  button_track_margin: 0.2
  safe_area:
    - 50
    - 20
    - 10
    - 25
  f0: 169.12836

